File: ios/FileDataLoader.swift
import Foundation

final class FileDataLoader: DataLoader {
  func load(from source: DataSource) async throws -> Data {
    guard case .file(let url) = source else {
      throw DataLoaderError.invalidSource
    }
    return try loadData(from: url)
  }

  func loadData(from url: URL) throws -> Data {
    guard FileManager.default.fileExists(atPath: url.path) else {
      throw DataLoaderError.fileNotFound(path: url.path)
    }
    return try Data(contentsOf: url)
  }

  func loadData(from path: String) throws -> Data {
    guard let url = URL(string: path), url.isFileURL else {
      throw DataLoaderError.invalidURL(path)
    }
    return try loadData(from: url)
  }
}

File: android/src/main/java/com/margelo/nitro/rive/FileDataLoader.kt
package com.margelo.nitro.rive

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import java.io.File

object FileDataLoader : DataLoader {
  override suspend fun load(source: DataSource): ByteArray {
    val fileSource = source as? DataSource.File
      ?: throw DataLoaderException.InvalidSource
    return loadBytes(fileSource.path)
  }

  suspend fun loadBytes(path: String): ByteArray = withContext(Dispatchers.IO) {
    val file = File(path)
    if (!file.exists()) {
      throw DataLoaderException.FileNotFound(path)
    }
    file.readBytes()
  }
}
